aptos-move/block-executor/src/types.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright © Aptos Foundation
+
+use aptos_types::transaction::BlockExecutableTransaction as Transaction;
+use std::{collections::HashSet, fmt};
+
+#[derive(Eq, Hash, PartialEq, Debug)]
+pub enum InputOutputKey<K, T, I> {
+    Resource(K),
+    Group(K, T),
+    DelayedField(I),
+}
+
+pub struct ReadWriteSummary<T: Transaction> {
+    reads: HashSet<InputOutputKey<T::Key, T::Tag, T::Identifier>>,
+    writes: HashSet<InputOutputKey<T::Key, T::Tag, T::Identifier>>,
+}
+
+impl<T: Transaction> ReadWriteSummary<T> {
+    pub fn new(
+        reads: HashSet<InputOutputKey<T::Key, T::Tag, T::Identifier>>,
+        writes: HashSet<InputOutputKey<T::Key, T::Tag, T::Identifier>>,
+    ) -> Self {
+        Self { reads, writes }
+    }
+
+    pub fn conflicts_with_previous(&self, previous: &Self) -> bool {
+        !self.reads.is_disjoint(&previous.writes)
+    }
+
+    pub fn collapse_resource_group_conflicts(self) -> Self {
```

Comment:
this is what handles groups in a backwards consistent way for now, right @igor-aptos ?